From 5f1c4ce6c4138aa735bf3db8531bcb58d38b8095 Mon Sep 17 00:00:00 2001
From: Sudip Mukherjee <sudipm.mukherjee@gmail.com>
Date: Wed, 6 Jan 2021 12:38:47 +0000
Subject: [PATCH] Add watchdog support

Signed-off-by: Sudip Mukherjee <sudipm.mukherjee@gmail.com>
---
 safety-app.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/safety-app.c b/safety-app.c
index dc9794c..f8db0e5 100644
--- a/safety-app.c
+++ b/safety-app.c
@@ -56,6 +56,9 @@ bool do_E2Echeck(unsigned char Message[6])
 
 int main(void)
 {
+	int wdt;
+	bool write_wdt = false;
+
 	// creating the communication pipe
 	const char *Pipe = "/tmp/safety-signal-source_to_safety-app";
 
@@ -66,11 +69,17 @@ int main(void)
 	//buffer for the message to be read
 	unsigned char Message[6];
 
+	wdt = open("/dev/watchdog0", O_WRONLY);
+	if (wdt >= 0)
+		write_wdt = true;
+
 	while (1) {
 		if (read(fd, Message, 6) > 0) {
 			printf("Received Message:%i %i %i %i %i %i\n", Message[0], Message[1], Message[2], Message[3], Message[4], Message[5]);
 			if (do_E2Echeck(Message)) {
 				printf("%s\n", Message);
+				if (write_wdt)
+					write(wdt, "1", 1);
 			} else {
 				printf("What a mess!, SAFESTATE\n");
 				// could just kill the qt app to get the safe state black screen
@@ -79,11 +88,13 @@ int main(void)
 				// system("killall afbd-cluster-gauges");
 				// Command QT app to do safe state display
 				system("cansend can0 021#0000000000000080");
+				write_wdt = false;
 			}
 			if (Message[0] == 0) {
 				// Signal source triggered safe state
 				printf("SAFESTATE (as commanded by signal source)\n");
 				system("cansend can0 021#0000000000000080");
+				write_wdt = false;
 			}
 		}
 		int sleeptime = 1000000;
@@ -94,5 +105,6 @@ int main(void)
 	}
 	//close file again
 	close(fd);
+	close(wdt);
 	return 0;
 }
-- 
2.20.1

