From 0d2fcc0cc7549defb0187a37cd0df5f5550eefa6 Mon Sep 17 00:00:00 2001
From: Sudip Mukherjee <sudipm.mukherjee@gmail.com>
Date: Wed, 13 Jan 2021 11:55:29 +0000
Subject: [PATCH] Add watchdog support

Signed-off-by: Sudip Mukherjee <sudipm.mukherjee@gmail.com>
---
 safety-app.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/safety-app.c b/safety-app.c
index dc9794c..3c2c2a7 100644
--- a/safety-app.c
+++ b/safety-app.c
@@ -56,6 +56,9 @@ bool do_E2Echeck(unsigned char Message[6])
 
 int main(void)
 {
+	int wdt;
+	bool write_wdt = false;
+
 	// creating the communication pipe
 	const char *Pipe = "/tmp/safety-signal-source_to_safety-app";
 
@@ -66,6 +69,10 @@ int main(void)
 	//buffer for the message to be read
 	unsigned char Message[6];
 
+	wdt = open("/dev/watchdog0", O_WRONLY);
+	if (wdt >= 0)
+		write_wdt = true;
+
 	while (1) {
 		if (read(fd, Message, 6) > 0) {
 			printf("Received Message:%i %i %i %i %i %i\n", Message[0], Message[1], Message[2], Message[3], Message[4], Message[5]);
@@ -79,12 +86,16 @@ int main(void)
 				// system("killall afbd-cluster-gauges");
 				// Command QT app to do safe state display
 				system("cansend can0 021#0000000000000080");
+				write_wdt = false;
 			}
 			if (Message[0] == 0) {
 				// Signal source triggered safe state
 				printf("SAFESTATE (as commanded by signal source)\n");
 				system("cansend can0 021#0000000000000080");
+				write_wdt = false;
 			}
+			if (write_wdt)
+				write(wdt, "1", 1);
 		}
 		int sleeptime = 1000000;
 
@@ -94,5 +105,6 @@ int main(void)
 	}
 	//close file again
 	close(fd);
+	close(wdt);
 	return 0;
 }
-- 
2.20.1

